stub/src/main/java/io/grpc/stub/ClientCalls.java

```diff
@@ -441,10 +448,33 @@ public void onMessage(RespT message) {
 
     @Override
     public void onClose(Status status, Metadata trailers) {
+      Throwable error = null;
       if (status.isOk()) {
-        observer.onCompleted();
+        boolean shouldCallOnCompleted = true;
+        if (!streamingResponse) {
+          shouldCallOnCompleted = false;
+          if (unaryMessage != null) {
+            try {
+              observer.onNext(unaryMessage);
+              shouldCallOnCompleted = true;
+            } catch (Throwable t) {
+              error = t;
+            }
+          } else {
+            error = Status.INTERNAL.withDescription("Response message is null for unary call")
+                .asRuntimeException();
+          }
+        }
+
+        if (shouldCallOnCompleted) {
+          observer.onCompleted();
```

Comment:
You can move this to an else at the end of the method, which makes it clear one of onError/onCompleted will be called. You also don't need `shouldCallOnCompleted` because you can use `error` instead.